
### Google PlayStore Android App Data. (10,000+ App Data)
- Context While many public datasets (on Kaggle and the like) provide Apple App Store data, there are not many counterpart datasets available for Google Play Store apps anywhere on the web. On digging deeper, I found out that iTunes App Store page deploys a nicely indexed appendix-like structure to allow for simple and easy web scraping. On the other hand, Google Play Store uses sophisticated modern-day techniques (like dynamic page load) using JQuery making scraping more challenging.

- Content Each app (row) has values for catergory, rating, size, and more.

- Acknowledgements This information is scraped from the Google Play Store. This app information would not be available without it.

Inspiration The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market!
- Backup repo: https://github.com/gauthamp10/Google-Playstore-Dataset

In [57]:
#importing Libraries:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express  as px




# import data

In [58]:
#importing Data
df=pd.read_csv('googleplaystore.csv')

In [59]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size     Installs  Type Price Content Rating  \
0      159   19M      10,000+  Free     0       Everyone   
1      967   14M     500,000+  Free     0       Everyone   
2    87510  8.7M   5,000,000+  Free     0       Everyone   
3   215644   25M  50,000,000+  Free     0           Teen   
4      967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10841 non-null  object 
 9   Genres          10840 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10839 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.1+ MB


In [61]:
df.describe()

Rating       Reviews
count  9367.000000  1.084100e+04
mean      4.191513  4.441119e+05
std       0.515735  2.927629e+06
min       1.000000  0.000000e+00
25%       4.000000  3.800000e+01
50%       4.300000  2.094000e+03
75%       4.500000  5.476800e+04
max       5.000000  7.815831e+07

## observations 

1. there are two numerical columns in the data set.
2. there are eleven catagorical or object type columns in the data set
3. there are also some null values in the data set.


#### We can check NULL values through the isnull function

In [62]:
df.isnull().sum()

App                  0
Category             1
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               1
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

there are: 

1474 null values in Rating column

1 null value in type column

1 null value in Genres column

8 null values in current version column

2 null values in Android version column

### Observations
1. There are 10841 rows and 13 columns in the dataset
2. The columns are of different data types
3. The columns in the datasets are:
'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',   'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',   'Android Ver'
4. There are some missing values in the dataset which we will read in details and deal later on in the notebook.
5. There are some columns which are of object data type but they should be of numeric data type, we will convert them later on in the notebook.
'Size', 'Installs', 'Price'

In [63]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [64]:
df['App'].unique()

array(['Photo Editor & Candy Camera & Grid & ScrapBook',
       'Coloring book moana',
       'U Launcher Lite – FREE Live Cool Themes, Hide Apps', ...,
       'Parkinson Exercices FR', 'The SCP Foundation DB fr nn5n',
       'iHoroscope - 2018 Daily Horoscope & Astrology'], dtype=object)

# Dealing with size column

In [65]:
df['Size'].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

the values are of object data type so we will convert them in numeric data type 

In [66]:
#checking total number of values with 'M' in them
df['Size'].str.contains('M').sum()

8830

In [67]:
#checking total number of values with 'K' in them
df['Size'].str.contains('k').sum()

316

In [68]:
# checking size that varies with device
df['Size'].str.contains('Varies with device').sum()

1695

as you see there are 8830 apps that are in MB's , 316 apps that are in KB's and 1695 apps varies with device

In [69]:
# convert whole size column to bytes
def convert_size(Size):
    if isinstance(Size, str):
        if "k" in Size:
            return float(Size.replace('k'," "))*1024
        elif 'M' in Size:
            return float(Size.replace('M'," "))*1024*1024
        elif 'Varies with device' in Size:
            return np.nan
    return Size
df['Size']=df['Size'].apply(convert_size)

In [70]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews        Size     Installs  Type Price Content Rating  \
0      159  19922944.0      10,000+  Free     0       Everyone   
1      967  14680064.0     500,000+  Free     0       Everyone   
2    87510   9122611.2   5,000,000+  Free     0       Everyone   
3   215644  26214400.0  50,000,000+  Free     0           Teen   
4      967   2936012.8     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [71]:
## convert size column to mega bytes
df['Size']=df['Size'].apply(lambda x: x/1024/1024 if x>1024 else x)
# change name of size column to size in MB
df.rename(columns={'Size':'Size(MB)'},inplace=True)   


In [72]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size(MB)     Installs  Type Price Content Rating  \
0      159      19.0      10,000+  Free     0       Everyone   
1      967      14.0     500,000+  Free     0       Everyone   
2    87510       8.7   5,000,000+  Free     0       Everyone   
3   215644      25.0  50,000,000+  Free     0           Teen   
4      967       2.8     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Mile stone 1 completed
We have convert the whole size column to the numeric data type

# Dealing with Installs column

In [73]:
df['Installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [74]:
df['Installs']=df['Installs'].str.replace('+','')
df['Installs']=df['Installs'].str.replace(',','')

In [75]:
df.head()

df['Installs']=df['Installs'].astype(int)

In [76]:
df['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

In [77]:
df['Price']=df['Price'].str.replace('$','')

In [78]:
df['Price']=df['Price'].astype(float)

In [79]:
df['Price'].dtype

dtype('float64')

In [80]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size(MB)  Installs  Type  Price Content Rating  \
0      159      19.0     10000  Free    0.0       Everyone   
1      967      14.0    500000  Free    0.0       Everyone   
2    87510       8.7   5000000  Free    0.0       Everyone   
3   215644      25.0  50000000  Free    0.0           Teen   
4      967       2.8    100000  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

#### Mile stone 2 completed
We have converted the installs column to numeric data type

In [81]:
df['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

In [82]:

# Fill null values in the 'Type' column with the mode
df['Type'] = df['Type'].fillna(df['Type'].mode()[0])

# Dealing with Content Rating column

In [83]:
#Observation:
# adult+mature category
# teen +everyone 10+ category
# everyone
# un-rated

In [84]:
df['Content Rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [85]:
df['Content Rating']=df['Content Rating'].str.replace('Mature 17+','Adults only 18+')

In [86]:

df['Content Rating']=df['Content Rating'].str.replace('Adults only 18+','Adults')

In [87]:
df['Content Rating']=df['Content Rating'].str.replace('Everyone 10+','Teen')

In [88]:
df['Content Rating'].unique()

array(['Everyone', 'Teen', 'Adults', 'Unrated'], dtype=object)

# we have categorize the content rating column in 
1. Everyone
2. Teen
3. Adults
4. Unrated

In [89]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size(MB)  Installs  Type  Price Content Rating  \
0      159      19.0     10000  Free    0.0       Everyone   
1      967      14.0    500000  Free    0.0       Everyone   
2    87510       8.7   5000000  Free    0.0       Everyone   
3   215644      25.0  50000000  Free    0.0           Teen   
4      967       2.8    100000  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Mile stone 3 
we have done dealing with content rating column

In [90]:
df.isnull().sum()

App                  0
Category             1
Rating            1474
Reviews              0
Size(MB)          1695
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               1
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

In [91]:
df.describe()

Rating       Reviews     Size(MB)      Installs         Price
count  9367.000000  1.084100e+04  9146.000000  1.084100e+04  10841.000000
mean      4.191513  4.441119e+05    21.514141  1.546291e+07      1.027273
std       0.515735  2.927629e+06    22.588679  8.502557e+07     15.948971
min       1.000000  0.000000e+00     0.008301  0.000000e+00      0.000000
25%       4.000000  3.800000e+01     4.900000  1.000000e+03      0.000000
50%       4.300000  2.094000e+03    13.000000  1.000000e+05      0.000000
75%       4.500000  5.476800e+04    30.000000  5.000000e+06      0.000000
max       5.000000  7.815831e+07   100.000000  1.000000e+09    400.000000

# Dealing With missing Values

In [92]:
#replace null values in rating column with mean
df['Rating'].fillna(df['Rating'].mean(),inplace=True)

C:\Users\nhash\AppData\Local\Temp\ipykernel_12260\3771142738.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [93]:
df['Category'].fillna(df['Category'].mode()[0],inplace=True)

C:\Users\nhash\AppData\Local\Temp\ipykernel_12260\1516221330.py:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [94]:
df['Genres'].fillna(df['Genres'].mode()[0],inplace=True)

C:\Users\nhash\AppData\Local\Temp\ipykernel_12260\4145008697.py:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [95]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size(MB)  Installs  Type  Price Content Rating  \
0      159      19.0     10000  Free    0.0       Everyone   
1      967      14.0    500000  Free    0.0       Everyone   
2    87510       8.7   5000000  Free    0.0       Everyone   
3   215644      25.0  50000000  Free    0.0           Teen   
4      967       2.8    100000  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Mile stone 4 completed 
We have done dealing with null  Values

# Dealing with duplicalte Values

In [96]:
#checking duplicates
df[df['App'].duplicated(keep=False)].sort_values(by='App')

App            Category  Rating  Reviews  \
1393           10 Best Foods for You  HEALTH_AND_FITNESS     4.0     2490   
1407           10 Best Foods for You  HEALTH_AND_FITNESS     4.0     2490   
2543      1800 Contacts - Lens Store             MEDICAL     4.7    23160   
2322      1800 Contacts - Lens Store             MEDICAL     4.7    23160   
2385      2017 EMRA Antibiotic Guide             MEDICAL     4.4       12   
...                              ...                 ...     ...      ...   
3202        trivago: Hotels & Travel    TRAVEL_AND_LOCAL     4.2   219848   
3118        trivago: Hotels & Travel    TRAVEL_AND_LOCAL     4.2   219848   
3103        trivago: Hotels & Travel    TRAVEL_AND_LOCAL     4.2   219848   
8291  wetter.com - Weather and Radar             WEATHER     4.2   189310   
3652  wetter.com - Weather and Radar             WEATHER     4.2   189313   

      Size(MB)  Installs  Type  Price Content Rating            Genres  \
1393       3.8    500000  Free   0.00           Teen  Health & Fitness   
1407       3.8    500000  Free   0.00           Teen  Health & Fitness   
2543      26.0   1000000  Free   0.00       Everyone           Medical   
2322      26.0   1000000  Free   0.00       Everyone           Medical   
2385       3.8      1000  Paid  16.99       Everyone           Medical   
...        ...       ...   ...    ...            ...               ...   
3202       NaN  50000000  Free   0.00       Everyone    Travel & Local   
3118       NaN  50000000  Free   0.00       Everyone    Travel & Local   
3103       NaN  50000000  Free   0.00       Everyone    Travel & Local   
8291      38.0  10000000  Free   0.00       Everyone           Weather   
3652      38.0  10000000  Free   0.00       Everyone           Weather   

           Last Updated         Current Ver         Android Ver  
1393  February 17, 2017                 1.9        2.3.3 and up  
1407  February 17, 2017                 1.9        2.3.3 and up  
2543      July 27, 2018               7.4.1          5.0 and up  
2322      July 27, 2018               7.4.1          5.0 and up  
2385   January 27, 2017               1.0.5        4.0.3 and up  
...                 ...                 ...                 ...  
3202     August 2, 2018  Varies with device  Varies with device  
3118     August 2, 2018  Varies with device  Varies with device  
3103     August 2, 2018  Varies with device  Varies with device  
8291     August 6, 2018  Varies with device  Varies with device  
3652     August 6, 2018  Varies with device  Varies with device  

[1979 rows x 13 columns]

In [97]:
# droping duplicates
df.drop_duplicates(inplace=True)

In [98]:
df.shape

(10358, 13)

In [99]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   Reviews  Size(MB)  Installs  Type  Price Content Rating  \
0      159      19.0     10000  Free    0.0       Everyone   
1      967      14.0    500000  Free    0.0       Everyone   
2    87510       8.7   5000000  Free    0.0       Everyone   
3   215644      25.0  50000000  Free    0.0           Teen   
4      967       2.8    100000  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Mile stone 5 completed
we have done dealing with duplicates

### So we have
1. deal with data types
2. deal with null values
3. deal with duplicates